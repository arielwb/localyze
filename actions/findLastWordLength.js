export const findLastWordLength = (string) => {
  const getLastWord = /(\w+)\s*$/
  const lastWord = getLastWord.exec(string)[1]

  return lastWord.length
}
