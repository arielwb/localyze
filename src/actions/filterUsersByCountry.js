export const sampleUsers = [
  { name: 'Maddy', country: 'Brazil' },
  { name: 'Nate', country: 'Netherlands' },
  { name: 'Cassie', country: 'Germany' },
  { name: 'Fezco', country: 'Norway' },
  { name: 'Rue', country: 'France' },
  { name: 'Jules', country: 'France' },
  { name: 'Elliot', country: 'Poland' },
  { name: 'Lexi', country: 'Poland' }
]

export const filterUsersByCountry = (setState, users, countryFilter) => {
  if (!countryFilter) {
    setState(users)
  } else {
    const filteredCountries = users.filter((user) => {
      const country = user.country.toLowerCase()
      const filter = countryFilter.toLowerCase()

      return country.includes(filter)
    })

    setState(filteredCountries)
  }
}
