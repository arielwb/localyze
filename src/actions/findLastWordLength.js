export const findLastWordLength = (string) => {
  if (string.length < 1 || string.length > 104) {
    return null
  }

  const getLastWord = /(\w+)\s*$/
  const lastWord = getLastWord.exec(string)[1]

  return lastWord.length
}
