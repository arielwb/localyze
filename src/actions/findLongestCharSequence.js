export const findLongestCharSequence = (word) => {
  let longestCharSequence = ''
  let currentCharSequence = ''
  const characters = [...word]

  characters.forEach(char => {
    if (!longestCharSequence) {
      longestCharSequence = char
      currentCharSequence = char

      return
    }

    currentCharSequence = currentCharSequence[0] !== char
      ? char
      : currentCharSequence + char

    if (currentCharSequence.length > longestCharSequence.length) {
      longestCharSequence = currentCharSequence
    }
  })

  return longestCharSequence
}
