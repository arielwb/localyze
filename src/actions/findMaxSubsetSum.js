const sum = (numbers) => numbers.reduce((currentSum, value) => currentSum + value, 0)

const findMaxNumberIndex = (numbers) => {
  let maxNumber = numbers[0]
  let maxNumberIndex = 0

  numbers.forEach((num, index) => {
    if (num > maxNumber) {
      maxNumber = num
      maxNumberIndex = index
    }
  })

  return maxNumberIndex
}

export const findMaxSubsetSum = (numbers) => {
  if (!numbers.length) return []

  const negativeNumbersIndex = []
  let prevIndex
  let prevSum = 0
  let startIndex = 0
  let endIndex = numbers.length - 1

  numbers.forEach((num, index) => num < 0 && negativeNumbersIndex.push(index))

  if (negativeNumbersIndex.length === numbers.length) {
    const maxNegativeNumberIndex = findMaxNumberIndex(numbers)

    return [maxNegativeNumberIndex, maxNegativeNumberIndex]
  }

  negativeNumbersIndex.forEach((negativeNumIndex) => {
    const range = numbers.slice(prevIndex, negativeNumIndex)
    const rangeSum = sum(range)

    if (prevIndex) {
      const positiveRangeSum = range.length === 1 ? rangeSum : rangeSum - numbers[prevIndex]
      const totalSum = prevSum + rangeSum

      if (positiveRangeSum > totalSum) {
        startIndex = prevIndex + 1
        prevSum = positiveRangeSum
        endIndex = negativeNumIndex - 1
      } else if (totalSum > prevSum) {
        prevSum = totalSum
        endIndex = negativeNumIndex - 1
      }
    } else {
      prevSum = rangeSum
      endIndex = negativeNumIndex - 1
    }

    prevIndex = negativeNumIndex
  })

  const lastNegativeIndex = negativeNumbersIndex.pop()

  if (lastNegativeIndex !== numbers.length - 1) {
    const lastRange = numbers.slice(lastNegativeIndex)
    const lastRangeSum = sum(lastRange)
    const lastPositiveRangeSum = lastRangeSum - numbers[lastNegativeIndex]

    if (lastPositiveRangeSum > prevSum) {
      startIndex = lastNegativeIndex + 1
      endIndex = numbers.length - 1
    } else if (prevSum + lastRangeSum > prevSum) {
      endIndex = numbers.length - 1
    }
  }

  return [startIndex, endIndex]
}
