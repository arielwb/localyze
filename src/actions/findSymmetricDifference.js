const getUniqueItems = (reference, set) => reference.filter(number => !set.has(number))

export const findSymmetricDifference = (sample1, sample2) => {
  if (sample1.length > 100 || sample2.length > 100) {
    return null
  }

  const set1 = new Set(sample1)
  const set2 = new Set(sample2)

  return [...getUniqueItems(sample1, set2), ...getUniqueItems(sample2, set1)]
}
