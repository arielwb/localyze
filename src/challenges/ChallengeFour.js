import React, { useState } from 'react'
import { Header, Input, OutputBox } from '../components'

const users = [
  { name: 'Maddy', country: 'Brazil' },
  { name: 'Nate', country: 'Netherlands' },
  { name: 'Cassie', country: 'Germany' },
  { name: 'Fezco', country: 'Norway' },
  { name: 'Rue', country: 'France' },
  { name: 'Jules', country: 'France' },
  { name: 'Elliot', country: 'Poland' },
  { name: 'Lexi', country: 'Poland' }
]

const filterUsers = (setState, value) => {
  if (!value) {
    setState(users)
  } else {
    const filteredCountries = users.filter((user) => {
      const country = user.country.toLowerCase()
      const search = value.toLowerCase()

      return country.includes(search)
    })

    setState(filteredCountries)
  }
}

export const ChallengeFour = () => {
  const buildUserList = (user, index) => (
    <li key={`${user.name}-${index}`}>
      <span>{user.name}</span>
      <span>{user.country}</span>
    </li>
  )

  const [userList, setUsers] = useState(users)

  return (
    <>
      <Header id={4}>
        Code a list of users with countries `type User = name: string; country: string` in React.js.
        You should be able to filter users from a specific country.
        What questions you would ask your product manager if you received this task?
      </Header>
      <Input
        type='search'
        name='challenge4'
        label='Filter by country'
        onChange={(event) => filterUsers(setUsers, event.target.value)}
      />
      <ul className='user-list'>{userList.map(buildUserList)}</ul>
      <OutputBox>
        <h2>Questions:</h2>
        <ul>
          <li>
            Where do I get the user data from? What is the url of service/api?
          </li>
          <li>
            Should we implement completion/suggestions of country names?
            Or maybe a dropdown of available countries?
          </li>
          <li>
            What will the user do with this information?
            Should we offer a friendly way to download/export the filtered data?
          </li>
          <li>
            Will this feature grow? Will this list have other actions like edit, delete?
          </li>
          <li>
            Do we have any guidelines on how the design should look like?
          </li>
        </ul>
      </OutputBox>
    </>
  )
}
