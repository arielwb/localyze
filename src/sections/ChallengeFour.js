import React, { useState } from 'react'
import { Header, Input, OutputBox } from '../components'
import { filterUsersByCountry, sampleUsers } from '../actions/filterUsersByCountry'

const buildUserList = (user, index) => (
  <li key={`${user.name}-${index}`}>
    <span>{user.name}</span>
    <span>{user.country}</span>
  </li>
)

export const ChallengeFour = () => {
  const [userList, setUsers] = useState(sampleUsers)

  return (
    <>
      <Header id={4}>
        Code a list of users with countries `type User = name: string; country: string` in React.js.
        You should be able to filter users from a specific country.
        What questions you would ask your product manager if you received this task?
      </Header>
      <Input
        type='search'
        name='challenge4'
        label='Filter by country'
        onChange={(event) => filterUsersByCountry(setUsers, event.target.value)}
      />
      <ul className='user-list'>{userList.map(buildUserList)}</ul>
      <OutputBox>
        <h2>Questions:</h2>
        <ul>
          <li>
            Where do I get the user data from? What is the url of service/api?
          </li>
          <li>
            Should we implement completion/suggestions of country names?
            Or maybe a dropdown of available countries?
          </li>
          <li>
            What will the user do with this information?
            Should we offer a friendly way to download/export the filtered data?
          </li>
          <li>
            Will this feature grow? Will this list have other actions like edit, delete?
          </li>
          <li>
            Do we have any guidelines on how the design should look like?
          </li>
        </ul>
      </OutputBox>
    </>
  )
}
