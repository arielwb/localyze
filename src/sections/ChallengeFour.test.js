import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react'
import { ChallengeFour } from './ChallengeFour'
import { filterUsersByCountry } from '../actions/filterUsersByCountry'

jest.mock('../actions/filterUsersByCountry', () => {
  return {
    __esModule: true,
    filterUsersByCountry: jest.fn(),
    sampleUsers: [{ name: 'bob', country: 'spain' }]
  }
})

describe('ChallengeFour', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match the snapshot', () => {
    const tree = renderer
      .create(<ChallengeFour />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call filterUsersByCountry function on change', () => {
    render(<ChallengeFour />)

    const input = screen.getByTestId('input')

    fireEvent.change(input, { target: { value: 'germany' } })

    expect(filterUsersByCountry).toHaveBeenCalled()
  })
})
