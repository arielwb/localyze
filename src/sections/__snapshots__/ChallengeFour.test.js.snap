// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengeFour should match the snapshot 1`] = `
Array [
  <h1>
    Challenge 
    4
  </h1>,
  <p>
    Code a list of users with countries \`type User = name: string; country: string\` in React.js. You should be able to filter users from a specific country. What questions you would ask your product manager if you received this task?
  </p>,
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="challenge4"
    >
      Filter by country
    </label>
    <input
      data-testid="input"
      name="challenge4"
      onChange={[Function]}
      type="search"
    />
  </form>,
  <ul
    className="user-list"
  >
    <li>
      <span>
        bob
      </span>
      <span>
        spain
      </span>
    </li>
  </ul>,
  <pre>
    <h2>
      Questions:
    </h2>
    <ul>
      <li>
        Where do I get the user data from? What is the url of service/api?
      </li>
      <li>
        Should we implement completion/suggestions of country names? Or maybe a dropdown of available countries?
      </li>
      <li>
        What will the user do with this information? Should we offer a friendly way to download/export the filtered data?
      </li>
      <li>
        Will this feature grow? Will this list have other actions like edit, delete?
      </li>
      <li>
        Do we have any guidelines on how the design should look like?
      </li>
    </ul>
  </pre>,
]
`;
